### PR TITLE
[pallet-revive] eth-prc fix geth diff

### PR DESCRIPTION
**Pull request by pgherveou**
* Add a bunch of differential tests to ensure that responses from eth-rpc matches the one from `geth`
  - These [tests](https://github.com/paritytech/polkadot-sdk/blob/pg/fix-geth-diff/substrate/frame/revive/rpc/examples/js/src/geth-diff.test.ts) are not run in CI for now but can be run locally with 
```bash
cd revive/rpc/examples/js
bun test
```

* EVM RPC server will not fail gas_estimation if no gas is specified, I updated pallet-revive to add an extra `skip_transfer` boolean check to replicate this behavior in our pallet

* `eth_transact` and `bare_eth_transact` api have been updated to use `GenericTransaction` directly as this is what is used by `eth_estimateGas` and `eth_call`

## TODO

- [ ]  Add tests the new `skip_transfer` flag

### CODE REVIEW

File: substrate/frame/revive/rpc/src/client.rs [hunk 1]
```diff
@@ -146,42 +170,47 @@ pub enum ClientError {
 	/// A [`codec::Error`] wrapper error.
 	#[error(transparent)]
 	CodecError(#[from] codec::Error),
-	/// The dry run failed.
-	#[error("Dry run failed: {0}")]
-	DryRunFailed(String),
 	/// Contract reverted
-	#[error("Execution reverted: {}", extract_revert_message(.0).unwrap_or_default())]
-	Reverted(Vec<u8>),
+	#[error("Contract reverted")]
```
Comment (pgherveou) [inline substrate/frame/revive/rpc/src/client.rs:174]:
```suggestion
	#[error("contract reverted")]
```

File: substrate/frame/revive/Cargo.toml [hunk 2]
```diff
@@ -65,6 +65,7 @@ pallet-revive-fixtures = { workspace = true, default-features = true }
 secp256k1 = { workspace = true, features = ["recovery"] }
 serde_json = { workspace = true }
 hex-literal = { workspace = true }
+env_logger = { workspace = true }
```
Comment (athei) [inline substrate/frame/revive/Cargo.toml:68]: Where is this used?

File: substrate/frame/revive/rpc/src/client.rs [hunk 3]
```diff
@@ -146,42 +170,47 @@ pub enum ClientError {
...
+	#[error("transactionFeePaid event not found")]
 	TxFeeNotFound,
 	/// The cache is empty.
-	#[error("Cache is empty")]
+	#[error("cache is empty")]
 	CacheEmpty,
 }
 
+const REVERT_CODE: i32 = 3;
 // TODO convert error code to https://eips.ethereum.org/EIPS/eip-1474#error-codes
```
Comment (athei) [inline substrate/frame/revive/rpc/src/client.rs:191]: Isn't that TODO fixed here?

File: substrate/frame/revive/src/exec.rs [hunk 4]
```diff
@@ -1073,6 +1082,7 @@ where
 					&frame.account_id,
 					frame.contract_info.get(&frame.account_id),
 					executable.code_info(),
+					self.skip_transfer,
```
Comment (athei) [inline substrate/frame/revive/src/exec.rs:1085]: Don't we need to mock the normal balance transfer further down in this function, too ( `Self::transfer_from_origin` ) ?

File: substrate/frame/revive/src/exec.rs [hunk 5]
```diff
@@ -562,6 +562,8 @@ pub struct Stack<'a, T: Config, E> {
 	debug_message: Option<&'a mut DebugBuffer>,
 	/// Transient storage used to store data, which is kept for the duration of a transaction.
 	transient_storage: TransientStorage<T>,
+	/// Whether or not actual transfer of funds should be performed.
+	skip_transfer: bool,
```
Comment (athei) [inline substrate/frame/revive/src/exec.rs:566]: Doc should contain explanation when and why this transfer is skipped.

File: substrate/frame/revive/src/lib.rs [hunk 6]
```diff
@@ -573,6 +573,8 @@ pub mod pallet {
 		AccountUnmapped,
 		/// Tried to map an account that is already mapped.
 		AccountAlreadyMapped,
+		/// The transaction used to dry-run a contract is invalid.
+		InvalidGenericTransaction,
```
Comment (xermicus) [inline substrate/frame/revive/src/lib.rs:577]: The comment suggest this error can only occur during dry-runs, could it also explain why?

**Review by xermicus — APPROVED** [review]
Just some nits

File: substrate/frame/revive/Cargo.toml (hunk 2 above)
Comment (pgherveou) [reply substrate/frame/revive/Cargo.toml:68]: no where I think I removed it from local tests

File: substrate/frame/revive/rpc/src/client.rs (hunk 3 above)
Comment (pgherveou) [reply substrate/frame/revive/rpc/src/client.rs:191]: Yep

File: substrate/frame/revive/Cargo.toml (hunk 2 above)
Comment (pgherveou) [reply substrate/frame/revive/Cargo.toml:68]: ah using it in some example will move it to dev-dependencies

File: substrate/frame/revive/rpc/revive_chain.metadata
[the remark this answers is not in the document]
Comment (pgherveou) [reply substrate/frame/revive/rpc/revive_chain.metadata:1]: a bit tricky to generate it you need to compile the kitchensink node with your last changes first

File: substrate/frame/revive/rpc/src/client.rs [hunk 7]
```diff
@@ -118,16 +118,40 @@ fn unwrap_call_err(err: &subxt::error::RpcError) -> Option<ErrorObjectOwned> {
...
 	let function_selector = exec_data.get(0..4)?;
 
-	// keccak256("Error(string)")
-	let expected_selector = [0x08, 0xC3, 0x79, 0xA0];
-	if function_selector != expected_selector {
-		return None;
-	}
+	match function_selector {
+		// assert(false)
+		[0x4E, 0x48, 0x7B, 0x71] => {
```
[the remark this answers is not in the document]
Comment (pgherveou) [reply substrate/frame/revive/rpc/src/client.rs:123]: Not sure, this is inspired by the go-ethereum source code

File: substrate/frame/revive/src/lib.rs (hunk 6 above)
Comment (pgherveou) [reply substrate/frame/revive/src/lib.rs:577]: look like this is an old error case not used anymore. :hocho:  it

File: substrate/frame/revive/src/exec.rs (hunk 4 above)
Comment (pgherveou) [reply substrate/frame/revive/src/exec.rs:1086]: No this one is fine, if you explicitly specify a value to transfer then your call should fail

https://github.com/paritytech/polkadot-sdk/blob/76e3fdaa26fdfec1d2ac71e25649f140f55d2156/substrate/frame/revive/rpc/examples/js/src/geth-diff.test.ts?plain=1#L132-L150

File: substrate/frame/revive/rpc/examples/js/contracts/ErrorTester.sol [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+pragma solidity ^0.8.0;
+
+contract ErrorTester {
+
+	// Payable function that can be used to test unsifficient funds errors
+    function valueMatch(uint value) public payable {
+		require(msg.value == value , "msg.value does not match value");
+    }
+
+	bool public state;
```
[the remark this answers is not in the document]
Comment (pgherveou) [reply substrate/frame/revive/rpc/examples/js/contracts/ErrorTester.sol]: addes solhint not sure if this is the most commonly used linter

File: substrate/frame/revive/rpc/src/client.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (pgherveou) [reply substrate/frame/revive/rpc/src/client.rs:123]: We don't depend on alloy, so not really worth adding a new dependency just for that for now

**Review by athei — APPROVED** [review]
I think you need to disable the tests written in Solidity for CI to pass. The 64bit PR broke them. Or you rebuilt them with a new revive.

**Comment by pgherveou** [comment]
> I think you need to disable the tests written in Solidity for CI to pass. The 64bit PR broke them. Or you rebuilt them with a new revive.

Didn't you do it in your PR already?

**Comment by athei** [comment]
> > I think you need to disable the tests written in Solidity for CI to pass. The 64bit PR broke them. Or you rebuilt them with a new revive.
> 
> Didn't you do it in your PR already?

I did. I thought you added new tests here that need to be disabled. But the CI error is from something else.

**Comment by pgherveou** [comment]
Ah i was going to add a few tests will do that in a follow up

**Comment by cmichi** [comment]
@pgherveou I think you might have overlooked in `fn bare_upload_code` changing the 

``
storage_deposit_limit: BalanceOf<T>,
``

to

```
storage_deposit_limit: DepositLimit<BalanceOf<T>>,
```

Here:
https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/revive/src/lib.rs#L1468

**Comment by pgherveou** [comment]
> @pgherveou I think you might have overlooked in `fn bare_upload_code` changing the
> 
> `storage_deposit_limit: BalanceOf<T>,`
> 
> to
> 
> ```
> storage_deposit_limit: DepositLimit<BalanceOf<T>>,
> ```
> 
> Here: https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/revive/src/lib.rs#L1468

Thanks @cmichi, this new type barely acts as a more meaningful replacement for `Option< DepositLimit<BalanceOf<T>>>`
so we don't really need it here

**Pull request merged**